lib/web_ui/lib/src/engine/history.dart

```diff
@@ -5,96 +5,262 @@
+  /// of this class.
+  @protected
+  Future<void> tearDown() => Future<void>.value();
+}
+
+/// A browser history class that creates a set of browser history entries to
+/// support browser backward and forward button natively.
+///
+/// This class pushes a browser history entry every time the framework reports
+/// a route change and sends a `pushRouteInformation` method call to the
+/// framework when the browser jumps to a specific browser history entry.
+///
+/// The web engine uses this class to manage its browser history when the
+/// framework uses a Router for routing.
+///
+/// See also:
+///
+/// * [SingleEntryBrowserHistory], which is used when the framework does not use
+///   a Router for routing.
+class MultiEntriesBrowserHistory extends BrowserHistory {
+  int _lastSeenSerialCount = 0;
+  int get _getCurrentSerialCount {
+    if (_hasSerialCount(currentState)) {
+      return currentState['serialCount'] as int;
+    }
+    return 0;
+  }
+
+  dynamic _tagWithSerialCount(dynamic originialState, int count) {
+    return <dynamic, dynamic> {
+      'serialCount': count,
+      'state': originialState,
+    };
+  }
+
+  bool _hasSerialCount(dynamic state) {
+    return state is Map && state['serialCount'] != null;
+  }
+
+  @override
+  void setRouteName(String? routeName, {dynamic? state}) {
+    if (locationStrategy != null && routeName != null) {
+      final int serialCount = _getCurrentSerialCount +1;
+      _lastSeenSerialCount = serialCount;
+      locationStrategy!.pushState(
+        _tagWithSerialCount(state, serialCount),
+        'flutter',
+        routeName
+      );
+    }
+  }
+
+  @override
+  void onPopState(covariant html.PopStateEvent event) {
+    assert(locationStrategy != null);
+    // May be a result of direct url access while the flutter application is
+    // already running.
+    if (!_hasSerialCount(event.state)) {
+      // In this case we assume this will be the next history entry from last
+      // seen entry.
+      locationStrategy!.replaceState(
+        _tagWithSerialCount(event.state, _lastSeenSerialCount + 1),
+        'flutter',
+        currentPath);
+    }
+    _lastSeenSerialCount = _getCurrentSerialCount;
+    if (window._onPlatformMessage != null) {
+      window.invokeOnPlatformMessage(
+        'flutter/navigation',
+        const JSONMethodCodec().encodeMethodCall(
+          MethodCall('pushRouteInformation', <dynamic, dynamic>{
+            'location': currentPath,
+            'state': event.state?['state'],
+          })
+        ),
+        (_) {},
+      );
+    }
+  }
+
+  @override
+  Future<void> setup() {
+    if (!_hasSerialCount(currentState)) {
+      locationStrategy!.replaceState(
+        _tagWithSerialCount(currentState, 0),
+        'flutter',
+        currentPath
+      );
     }
     return Future<void>.value();
   }
 
+  @override
+  Future<void> tearDown() async {
+    // Restores the html browser history.
+    if (_hasSerialCount(currentState)) {
```

Comment:
If everything goes well, this condition should always be true right? If it's false, that means someone other than us is modifying history state. What do you think about doing an `assert` instead?